<a href="https://colab.research.google.com/github/JerryYu0815/Jie_Yu_CSC_448_AI/blob/main/CSC448_Assignment_4_Jie_Yu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1006]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd
from sklearn import neighbors, datasets, svm
import matplotlib.pyplot as plt
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier


# Part 1: Load the dataset

In [1007]:
# Load the dataset (load remotely, not locally)
iris = datasets.load_iris()
# Output the first 15 rows of the data
df = pd.DataFrame(iris.data, iris.target, columns=iris.feature_names)
print(df.head(15))

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
0                4.9               3.0                1.4               0.2
0                4.7               3.2                1.3               0.2
0                4.6               3.1                1.5               0.2
0                5.0               3.6                1.4               0.2
0                5.4               3.9                1.7               0.4
0                4.6               3.4                1.4               0.3
0                5.0               3.4                1.5               0.2
0                4.4               2.9                1.4               0.2
0                4.9               3.1                1.5               0.1
0                5.4               3.7                1.5               0.2
0                4.8               3.4                1.6               0.2
0           

In [1008]:
# Display a summary of the table information (number of datapoints, etc.)
print('\nTable information Summary:')
print('\nThe table description:\n',iris.DESCR)


Table information Summary:

The table description:
 .. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
 

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?
### The iris datasets contains 3 target class of flowers - setosa, versicolor, and virginica. Each target has 4 features - sepal length (cm), sepal width (cm), petal length (cm), petal width (cm). 
### Features -  sepal length (cm), sepal width (cm), petal length (cm), petal width (cm).
### Labels   -  0 for setosa, 1 for versicolor, and 2 for virginica.

# Part 2: Split the dataset into train and test

In [1009]:
# Take the dataset and split it into our features (X) and label (y)
X,y = load_iris(return_X_y = True)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=2)
print('Length of each set is:')
print('X_train: ',len(X_train))
print('X_test :',len(X_test))
print('y_train: ',len(y_train))
print('y_test: ',len(y_test))

Length of each set is:
X_train:  135
X_test : 15
y_train:  135
y_test:  15


# Part 3: Logistic Regression

In [1010]:
# i. Use sklearn to train a LogisticRegression model on the training set
clf = LogisticRegression(random_state=1).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print('First row in X_test[0]:', X_test[0])
print('First row in y_test[0]:', y_test[0])
pred = clf.predict(np.array([X_test[0]]))
print('The predicted result for X_test[0]= ',pred)
print('The actual reuslt y_test[0]= ',y_test[0])

print('\nFor X_test[0], the probabilities for each possible class[0 1 2] are:')
print(clf.predict_proba(X_test[:1, :]))

# iii. Report on the score for Logistic regression model, what does the score measure?
score = clf.score(X_train, y_train)
print('\nscore is %.3f'%score,',which means that the probability for our model to make correct prediction is %.2f'%(score*100)+'%') 

# iv. Extract the coefficents and intercepts for the boundary line(s)
coef = clf.coef_
intercept = clf.intercept_
print('\ncoef = ', coef)
print('intercept= ', intercept)


First row in X_test[0]: [4.6 3.4 1.4 0.3]
First row in y_test[0]: 0
The predicted result for X_test[0]=  [0]
The actual reuslt y_test[0]=  0

For X_test[0], the probabilities for each possible class[0 1 2] are:
[[9.83462916e-01 1.65370402e-02 4.36997045e-08]]

score is 0.970 ,which means that the probability for our model to make correct prediction is 97.04%

coef =  [[-0.39122448  0.93904954 -2.43765333 -1.02704598]
 [ 0.53482607 -0.39022414 -0.18049269 -0.97676208]
 [-0.14360159 -0.5488254   2.61814602  2.00380806]]
intercept=  [  9.3326014    2.33171215 -11.66431355]


## Score is 0.97 ,which means that the probability for our model to make correct prediction is 97%

# Part 4: Support Vector Machine

In [1011]:
# i. Use sklearn to train a Support Vector Classifier on the training set
clf = svm.SVC(kernel = 'linear',gamma= 'auto',probability=True)
clf.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
print('X_test[0]:', X_test[0])
print('y_test[0]:', y_test[0])
pred = clf.predict(np.array([X_test[0]]))
print('The predicted result for X_test[0]= ',pred)
print('\nFor X_test[0], the probabilities for each possible class[0 1 2] are:')
print(clf.predict_proba(X_test[:1, :]))

# iii. Report on the score for the SVM, what does the score measure?
score = clf.score(X_train, y_train)
print('\nscore is %.4f'%score,',which means that the probability for our model to make correct prediction is %.2f'%(score*100)+'%') 

X_test[0]: [4.6 3.4 1.4 0.3]
y_test[0]: 0
The predicted result for X_test[0]=  [0]

For X_test[0], the probabilities for each possible class[0 1 2] are:
[[0.96445484 0.0234517  0.01209346]]

score is 0.9926 ,which means that the probability for our model to make correct prediction is 99.26%


## The prediction result for X_test[0] = 0 and the the actual reuslt y_test[0] = 0. For the probablities of each possible class, we can see that the first class(0) has highest probability which is 0.958 which means that the prediction is accurate.
## Score is 0.963 ,which means that the probability for our model to make correct prediction is 96.3%

# Part 5: Neural Network

In [1012]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
clf = MLPClassifier(solver='lbfgs',hidden_layer_sizes=(3,3), alpha=1e-5, random_state=1).fit(X_train,y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

print('X_test[0]:', X_test[0])
print('y_test[0]:', y_test[0])
pred = clf.predict(np.array([X_test[0]]))
print('The predicted result for X_test[0]= ',pred)

print('\nFor X_test[0], the probabilities for each possible class[0 1 2] are:')
print(clf.predict_proba(X_test[:1, :]))
# iii. Report on the score for the Neural Network, what does the score measure?
score = clf.score(X_train, y_train)
print('\nOption: lbfgs - converge faster and perform better for small datasets')
print('score = ', score) 


# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
clf = MLPClassifier(solver='sgd',hidden_layer_sizes=(3,3),alpha=1e-5,random_state=1,max_iter = 3000).fit(X_train,y_train)

score = clf.score(X_train, y_train)
print('Option: sgd - perform well generally')
print('score = ', score)

clf = MLPClassifier(solver='adam',hidden_layer_sizes=(3,3),alpha=1e-5, random_state=1,max_iter = 3000).fit(X_train,y_train)
print('Option: adam - works pretty well on relatively large datasets')
score = clf.score(X_train, y_train)
print('score = ', score)


X_test[0]: [4.6 3.4 1.4 0.3]
y_test[0]: 0
The predicted result for X_test[0]=  [0]

For X_test[0], the probabilities for each possible class[0 1 2] are:
[[1.00000000e+000 5.26761204e-054 7.22758979e-158]]

Option: lbfgs - converge faster and perform better for small datasets
score =  0.9925925925925926
Option: sgd - perform well generally
score =  0.9851851851851852
Option: adam - works pretty well on relatively large datasets
score =  0.9777777777777777


## The prediction result for X_test[0] = 0 and the the actual reuslt y_test[0] = 0. For the probablities of each possible class, we can see that the first class(0) has highest probability which is 1.00 which means that the prediction is accurate.

## For lbfgs, Score is 0.993. ---- Highest Score
## For sgd, Score is 0.985.
## For adam, Score is 0.977.


# Part 6: K-Nearest Neighbors

In [1013]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
clf = KNeighborsClassifier(n_neighbors=3)
clf.fit(X_train, y_train)
pred = clf.predict(X_test)

# ii. For a sample datapoint, predict the probabilities for each possible class
print('X_test[0]:', X_test[0])
print('y_test[0]:', y_test[0])
pred = clf.predict(np.array([X_test[0]]))
print('The predicted result for X_test[0]= ',pred)
print('\nFor X_test[0], the probabilities for each possible class[0 1 2] are:')
print(clf.predict_proba(X_test[:1, :]))

# iii. Report on the score for kNN, what does the score measure?
score = clf.score(X_train, y_train)
print('\nscore is %.3f'%score,',which means that the probability for our model to make correct prediction is %.1f'%(score*100)+'%') 

X_test[0]: [4.6 3.4 1.4 0.3]
y_test[0]: 0
The predicted result for X_test[0]=  [0]

For X_test[0], the probabilities for each possible class[0 1 2] are:
[[1. 0. 0.]]

score is 0.956 ,which means that the probability for our model to make correct prediction is 95.6%


## Score is 0.963 ,which means that the probability for our model to make correct prediction is 96.3%

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

## According to performance of my models, it appears that Neural Network models using 'lbfgs' and Support Vector Machine using linear kernal makes the most accurate predictions on iris dataset. It looks right because Neural network models are good at classifying when we have a labeled dataset like iris and SVM is good for finding an optimal boundary between outputs.  While K-NN is sensitive to noisy data and outliears and so does logistic regression. Hence we might need to modify the data before traning in order to get accurate traning results for K-NN and Logistic regression.   